README.md
# apple-tv-converter

Command line utility to convert media to a format playable on the AppleTV. Supports embedding subtitles and sets metadata according to the media file.
Now, it also supports automatically downloading subtitles from [opensubtitles.org](http://www.opensubtitles.org) and, in the same process, infering the movie's IMDB id for metadata.

## On Mac OSX

- Uses [ffmpeg](http://ffmpeg.org) to encode video and audio, as well as extracting embedded subtitles
- Uses [Subler](http://code.google.com/p/subler/)'s command line interface to add the subtitles and set the iTunes metadata tags

# Command line usage

``` bash
Usage: apple-tv-converter [options] [file]
       [file] must be provided unless the -d (--dir) switch is present.
    -i, --id id                      Set a specific id for fetching metadata from online services
        --imdb_id id                 Set a specific id for fetching metadata from IMDB
        --tvdb_id id                 Set a specific id for fetching metadata from TheTVDB
        --tmdb_id id                 Set a specific id for fetching metadata from TheMovieDB
    -l, --languages eng,por,...      Only keep audio and subtitles in the specified languages
    -d, --dir DIRECTORY              Process all files in DIRECTORY recursively
        --itunes                     Add processed file to iTunes library, if it isn't present yet
        --os [USERNAME:PASSWORD]     Download subtitles and infer IMDB ID from opensubtitles.org
        --plex                       Rename file(s) to Plex Media Server recommended format

        --no-transcoding             Don't transcode video or audio
        --no-subtitles               Don't add subtitles
        --no-metadata                Don't add metadata (implies --no-online-metadata)
        --no-online-metadata         Don't fetch metadata from online services (IMDB or TheTVDB)
        --no-interactive             Perform all operations without user intervention, using sensible defaults
        --no-cleanup                 Don't cleanup the source files after processing

Advanced options:
        --use-absolute-numbering     Use absolute numbering for TV Show episodes (specially useful for cartoons)
        --episode-number-padding NUMBER
                                     Set the episode number padding length (ie, 3 for 001, 002, etc.)
    -s, --season NUMBER              Set the season number for TV Shows in case folder/file naming scheme doesn't contain right season
    -e, --episode NUMBER             Set the episode number for TV Shows in case folder/file naming scheme doesn't contain right episode number
        --width NUMBER               Resize the video to the specified width. If used with --height, can result in a different aspect ratio
        --height NUMBER              Resize the video to the specified height. If used with --width, can result in a different aspect ratio

Other options:
    -f, --ffmpeg LOCATION            Set path to ffmpeg binary

DEPRECATED options:
        --imdb                       Gather data from IMDB (optionally specifying movie id)

Common options:
    -h, --help                       Show this message
        --version                    Show version
```

## Remarks

### Subtitles

- External subtitles must be in SubRip (srt) format and have the same name as the file, optionally appending the language's code ie, for movie `/Home Movie 1/movie.mkv`, will load subtitles from file `/Home Movie 1/movie.srt` or `/Home Movie 1/movie.<language>.mkv`
    - Subtitles with appended language code will have the correct language set in metadata.
        - Language codes should be [ISO 639-3](http://www.iso.org/iso/home/standards/language_codes.htm) codes (mostly, I haven't quite figured all the languages yet =)

    Example:
    For the movie file `/Home Movie 1/movie.mkv`, the following subtitle files are loaded:

    - `/Home Movie 1/movie.srt` (language 'Unknown')
    - `/Home Movie 1/movie.eng.srt` (language 'English')
    - `/Home Movie 1/movie.2.por.srt` (language 'Portuguese'. Note: the _2_ is an mkvinfo id when extracted from an MKV movie)

### Metadata

Metadata can be obtained automatically from [themoviedb.org](http://www.themoviedb.org) (for movies) or from [TheTVDB.com](http://www.thetvdb.com) (for TV show episodes).
Both fall back to [IMDB.com](http://www.imdb.com) for necessary information. Most metadata will be filled, including the file's artwork, so it displays a nice image on iTunes library.

#### Fallbacks

If the data can't be found on neither site, there is a fallback for how metadata is set:
- The file metadata name is set according to the file's directory name, ie, the file `/Home Movie 1/movie.mkv` is converted to `/Home Movie 1/movie.mp4` and the name on metadata is set as `Home Movie 1`.
    - An exception to this is when the directory name has the format `Season XX`. In this situation the name is obtained from the 'grandparent' folder, ie, the file `/Home Movie 1/Season 1/movie.mp4` wil have the name on metadata set to `Home Movie 1`.
- For TV Show episodes, the metadata name is a concatenation of the directory (removing a reference to the season number, in the format Sxx) with the season and episode number, ie, both `/TV Show S01/tv.show.s01e01.mkv` and `/TV Show/Season 1/tv.show.s01e01.mkv` will have, after conversion, the name `TV Show S01E01`.
    Episode and season information is captured from the file name, in the formats S01E01 or 1x01 (case insensitive).
- The file genre is set from the directory name, for tv shows (ie, `/TV Show S01/tv.show.s01e01.mkv` will have the genre `TV Show`), and according to the movie's size (ie, `1080p Movies`, `720p Movies`, and `XviD Movies`).

### Plex Media Server

For Plex Media Server users, you can pass the command-line option `--plex` to automatically rename the file after conversion following the recommended file name convention. For files with multiple TV show episodes, the first and last episodes are identified in the file name (if it was possible to infer from the original filename).

### Other remarks

After conversion, `apple-tv-converter` will create a file named `.apple-tv-converter.data` on the base directory of the file containing some information (metadata service ids, etc.) that can be useful for subsequent processing.

## Thanks

* Subtitles service powered by [www.OpenSubtitles.org](http://www.opensubtitles.org)
* Movie metadata service powered by [themoviedb.org](http://www.themoviedb.org) (**This product uses the TMDb API but is not endorsed or certified by TMDb.**)
* TV Show metadata service powered by [www.thetvdb.com](http://www.thetvdb.com)
* Movie and TV show metadata service also powered by [www.imdb.com](http://www.imdb.com)

# TODO

- Windows implementation
- Tests!

CHANGELOG
v0.7.3
------
* Fixed an error that occured when a movie's metadata from TheMovieDB didn't provide a release date

v0.7.2
------
* Small fix on the command line help

v0.7.1
------
* Fixed `undefined method 'empty?' for Fixnum` on TVDB fetcher
* Fixed the poster for TV shows not being correctly set
* Removed forgotten debug messages

v0.7.0
------
* Added support for fetching metadata from themoviedb.org
* Performed some reengineering of the metadata services, in order to simplify adding more providers
* Fixed some bugs

v0.6.2
------
* Added the missing VERSION file

v0.6.1 (yanked)
------
* Fixed the display of the AKA names for movies
* Updated some gem dependencies

v0.6.0
------
* Added command line option to allow resizing the converted video
* Enchanced the --os option to allow supplying an username:password combination
* Other fixes

v0.5.9
------
* Fixed #5 Command line options now correctly override the data file (thanks @shir)

v0.5.8
------
* Refactored the filename parsing into a separate file and added parsing of "X_of_Y" and "X of Y" filenames (thanks @shir for pull request #4)

v0.5.7
------
* Added "--season" option (thanks @shir)
* Added "--episode" option
* Added "--imdb_id" option
* Added "--tvdb_id" option
* Updated "--id" option to set both IMDB and TheTVDB ids
* Fixed #1 Couldn't parse filename

v0.5.6
------
* Fixed "libfaac: Specified channel layout '2.1' is not supported" error

v0.5.5
------
* Fixed gem version requirements

v0.5.4
------
* Convert DTS 6.1 audio to 6 channels when transcoding

v0.5.3
------
* Some transcoding fixes
* Enable defining a padding length for the episode number via the data file (for example, for shows with more than 100 episodes per season)
* For season 1 episodes, fetch TheTVDB data first from the absolute episode number (as most anime shows don’t have the concept of seasons), and only after from the 'correct' season and episode number

v0.5.2
------
* Set sort metadata fields for TV Show episodes

v0.5.1
------
* Added option for renaming a file to the format recommended by Plex
* Now fetches metadata for TV show episodes from TheTVDB.com instead of IMDB
* When fetching metadata by name, and several records are found, the user now has the choice of which to use
* Correctly set Media Kind metadata tag for both TV show episodes and Movies
* Some other fixes

v0.4.4
------
* Support Ruby v1.8.7
* Fetch IMDB metadata for TV show episodes as well

v0.4.1
------
* Automatic download of subtitles via opensubtitles.org
* Automatic infer IMDB id from opensubtitles.org

v0.3.3
------
* Fixed SublerCLI call

v0.3.2
------
* Fixed AC3 conversion failing in some cases

v0.3.1
------
* Support for IMDB ids starting with 0

v0.3.0
------
* Removed dependency on MKV gem and MKVToolnix
* Added support for converting OGG videos
* Other fixes

v0.2.0
------
* Add to iTunes
* Filter audio and subtitles by language
* Some fixes

v0.1.1
------
* Better output

v0.1.0
------
* Added support for supplying directories to command line utility
* Better quality encoding of audio and video
* Create movies name .mp4 instead of .m4v

v0.0.2
------
* Added command line utility to convert media
* Do not clean up all subtitles in the folder after processing one file, ensure that only the current file's are deleted

v0.0.1
------
* Mac OSX implementation

lib/apple_tv_converter/command_line.rb
module AppleTvConverter
  class CommandLine
    def initialize(*args)
      @skip_subtitles = false
      @skip_metadata = false
      @skip_cleanup = false

      begin
        options = parse_arguments(args)

        media_objects = options.media

        converter = AppleTvConverter::MediaConverter.new(options)

        media_objects.sort { |a, b| a.original_filename <=> b.original_filename }.each_with_index do |media, index|
          puts "---[ Processing file #{index + 1} of #{media_objects.length}: #{File.basename(media.original_filename)} ]----------------"
          converter.process_media media
        end
      rescue ArgumentError => e
        puts "Error: #{e.message}"
      rescue => e
        puts "Error: #{e.message}"
        puts e.backtrace
      end
    end

    private

      def parse_arguments(arguments)
        require 'optparse'
        require 'optparse/time'
        require 'ostruct'

        id_switch = 0

        options = OpenStruct.new
        options.skip_transcoding = false
        options.skip_subtitles = false
        options.skip_metadata = false
        options.skip_cleanup = false
        options.add_to_itunes = false
        options.skip_online_metadata = false
        options.plex_format = false
        options.interactive = true
        options.imdb_id = nil
        options.tvdb_id = nil
        options.tmdb_id = nil
        options.use_absolute_numbering = false
        options.episode_number_padding = nil
        options.languages = []
        options.media = []
        options.season = nil
        options.episode = nil
        options.width = -1
        options.height = -1

        opts = OptionParser.new do |opts|
          opts.banner = "Usage: apple-tv-converter [options] [file]\n" +
                        "       [file] must be provided unless the -d (--dir) switch is present.\n"

          opts.on('-i', '--id id', "Set a specific id for fetching metadata from online services") do |id|
            raise ArgumentError.new("Can't supply both --id and --imdb_id, --tvdb_id, or --tmdb_id at the same time!") if id_switch > 0

            id_switch = 7
            options.imdb_id = id
            options.tvdb_id = id
            options.tmdb_id = id
          end

          opts.on('--imdb_id id', "Set a specific id for fetching metadata from IMDB") do |id|
            raise ArgumentError.new("Can't supply both --id and --imdb_id, --tvdb_id, or --tmdb_id at the same time!") if id_switch & 1 > 0

            id_switch |= 1
            options.imdb_id = id
          end

          opts.on('--tvdb_id id', "Set a specific id for fetching metadata from TheTVDB") do |id|
            raise ArgumentError.new("Can't supply both --id and --imdb_id, --tvdb_id, or --tmdb_id at the same time!") if id_switch & 2 > 0

            id_switch |= 2
            options.tvdb_id = id
          end

          opts.on('--tmdb_id id', "Set a specific id for fetching metadata from TheMovieDB") do |id|
            raise ArgumentError.new("Can't supply both --id and --imdb_id, --tvdb_id, or --tmdb_id at the same time!") if id_switch & 4 > 0

            id_switch |= 4
            options.tmdb_id = id
          end

          opts.on('-l', '--languages eng,por,...', Array, "Only keep audio and subtitles in the specified languages") do |languages|
            options.languages.push *languages
            # If filtering by languages, always include the undetermined language
            options.languages.push 'und' unless options.languages.include?('und')
            options.languages.map!(&:to_sym)
          end

          opts.on('-d', '--dir DIRECTORY', 'Process all files in DIRECTORY recursively') do |dir|
            raise ArgumentError.new("Path not found: #{dir}") unless File.exists?(dir)
            raise ArgumentError.new("Path is not a directory: #{dir}") unless File.directory?(dir)

            found_files = Dir[File.join(dir, '**', '*')].delete_if do |f|
              # Skip files with subtitle or ignored extensions, or directories
              File.directory?(f) || [AppleTvConverter::Media.subtitle_extensions + AppleTvConverter::Media.ignored_extensions].flatten.include?(File.extname(f).gsub(/\./, '').downcase)
            end

            options.media.push *(found_files.map do |file|
              parse_filename(file)
            end.compact)
          end

          opts.on('--itunes', "Add processed file to iTunes library, if it isn't present yet") do |i|
            options.add_to_itunes = true
          end

          opts.on('--os [USERNAME:PASSWORD]', "Download subtitles and infer IMDB ID from opensubtitles.org") do |username_password|
            options.download_subtitles = true
            if username_password =~ /^(.*?)\:(.*?)/
              options.download_subtitles_username = $1 if username_password =~ /^(.+?)\:.+$/
              options.download_subtitles_password = $1 if username_password =~ /^.+?\:(.+)$/
            end

            options.download_subtitles_username = nil if options.download_subtitles_username == ''
            options.download_subtitles_password = nil if options.download_subtitles_password == ''
          end

          opts.on('--plex', 'Rename file(s) to Plex Media Server recommended format') do
            options.plex_format = true
            options.skip_online_metadata = false
          end

          opts.separator ""

          opts.on('--no-transcoding', "Don't transcode video or audio") do |v|
            options.skip_transcoding = true
          end

          opts.on('--no-subtitles', "Don't add subtitles") do |v|
            options.skip_subtitles = true
          end

          opts.on('--no-metadata', "Don't add metadata (implies --no-online-metadata)") do |m|
            options.skip_metadata = true
          end

          opts.on('--no-online-metadata', "Don't fetch metadata from online services (IMDB or TheTVDB)") do |m|
            options.skip_online_metadata = true
          end

          opts.on('--no-interactive', "Perform all operations without user intervention, using sensible defaults") do |m|
            options.interactive = false
          end

          opts.on('--no-cleanup', "Don't cleanup the source files after processing") do |c|
            options.skip_cleanup = true
          end

          opts.separator ""
          opts.separator "Advanced options:"

          opts.on('--use-absolute-numbering', 'Use absolute numbering for TV Show episodes (specially useful for cartoons)') do |f|
            options.use_absolute_numbering = true
          end

          opts.on('--episode-number-padding NUMBER', 'Set the episode number padding length (ie, 3 for 001, 002, etc.)') do |i|
            options.episode_number_padding = i.to_i
          end

          opts.on('-s', '--season NUMBER', 'Set the season number for TV Shows in case folder/file naming scheme doesn\'t contain right season') do |i|
            options.season = i.to_i
          end

          opts.on('-e', '--episode NUMBER', 'Set the episode number for TV Shows in case folder/file naming scheme doesn\'t contain right episode number') do |i|
            options.episode = i.to_i
          end

          opts.on('--width NUMBER', 'Resize the video to the specified width. If used with --height, can result in a different aspect ratio') do |i|
            options.width = i.to_i
            options.width -= options.width % 2 # Ensure it's always a even number
          end

          opts.on('--height NUMBER', 'Resize the video to the specified height. If used with --width, can result in a different aspect ratio') do |i|
            options.height = i.to_i
            options.height -= options.height % 2 # Ensure it's always a even number
          end

          opts.separator ""
          opts.separator "Other options:"

          opts.on('-f', '--ffmpeg LOCATION', 'Set path to ffmpeg binary') do |f|
            FFMPEG.ffmpeg_binary = f
          end

          opts.separator ""
          opts.separator "DEPRECATED options:"

          opts.on('--imdb', "Gather data from IMDB (optionally specifying movie id)") do
            puts "Warning: Switch --imdb is DEPRECATED, and will be removed in a future version. It is now activated by default"
            puts "         If you want to specify an id, please use the switch --id."
          end

          opts.separator ""
          opts.separator "Common options:"

          # No argument, shows at tail.  This will print an options summary.
          # Try it and see!
          opts.on_tail("-h", "--help", "Show this message") do
            puts opts
            exit
          end

          # Another typical switch to print the version.
          opts.on_tail("--version", "Show version") do
            puts AppleTvConverter::VERSION
            exit
          end
        end

        opts.parse! arguments
        options.media.push *(arguments.map { |file| parse_filename(file) }.compact)

        raise ArgumentError.new("No media file supplied") unless options.media.any?

        return options
      end

      def parse_filename(file)
        begin
          return nil unless FFMPEG::Movie.new(file).valid?

          begin
            e = AppleTvConverter::Media.new

            parser = FilenameParser.new(file)

            e.show        = parser.tvshow_name
            e.season      = parser.season_number
            e.number      = parser.episode_number
            e.last_number = parser.last_episode_number

            e.original_filename = file

            return e
          rescue => exc
            puts "File.expand_path(file): #{File.expand_path(file)}"
            puts "Couldn't parse filename, skipping: #{File.basename(file)}"
            puts "Reason: #{exc.respond_to?(:message) ? exc.message : exc}"
            puts "In: #{exc.backtrace.join("\n    ")}"

            return nil
          end
        rescue Errno::ENOENT
          puts "File not found: #{file}"
        rescue Exception => e
          puts e

          exit!
        end

      end
  end
end